orttraining/orttraining/core/optimizer/compute_optimizer/padding_elimination.cc

```diff
@@ -0,0 +1,568 @@
+                                                                        nullptr)};
+  Node* new_reshape_node = onnxruntime::optimizer::compute_optimizer::InsertIntermediateNodeOnDestInput(
+      graph, node,
+      in_index,
+      0,
+      0,
+      graph.GenerateNodeName("RecoverShape"),
+      "Reshape",
+      "Reshape node to recover invalid tokens.",
+      reshape_input_args,
+      reshape_output_args,
+      {},
+      kOnnxDomain,
+      logger);
+  new_reshape_node->SetExecutionProviderType(node.GetExecutionProviderType());
+  return new_reshape_node->MutableOutputDefs()[0];
+}
+
+// Iterate the subgraph beginning from the start_node, and put all node args contained in the subgraph into
+// candidate_node_args. Also put all input nodes and output nodes of the subgraph into candidate_input and
+// candidate_output respectively.
+void IterateSubgraphFromNode(Graph& graph,
+                                  Node* start_node,
+                                  std::unordered_set<NodeArg*>& candidate_node_args,
+                                  std::unordered_set<Node*>& candidate_input,
+                                  std::unordered_set<Node*>& candidate_output,
+                                  const logging::Logger& logger) {
+  std::queue<Node*> to_visit;
+  std::unordered_set<Node*> visited;
+  PushAllOutputNode(graph, to_visit, start_node, visited);
+  while (!to_visit.empty()) {
+    Node* cur = to_visit.front();
+    to_visit.pop();
+    visited.insert(cur);
+    if (graph_utils::IsSupportedOptypeVersionAndDomain(*cur, "Add", {7, 13, 14})) {
+      ORT_ENFORCE(candidate_node_args.find(cur->MutableInputDefs()[0]) != candidate_node_args.end() ||
+                  candidate_node_args.find(cur->MutableInputDefs()[1]) != candidate_node_args.end());
+      NodeArg* arg_in_subgraph = nullptr;
+      NodeArg* arg_not_in_subgraph = nullptr;
+      if (candidate_node_args.find(cur->MutableInputDefs()[0]) != candidate_node_args.end()) {
+        arg_in_subgraph = cur->MutableInputDefs()[0];
+        arg_not_in_subgraph = cur->MutableInputDefs()[1];
+      } else if (candidate_node_args.find(cur->MutableInputDefs()[1]) != candidate_node_args.end()) {
+        arg_in_subgraph = cur->MutableInputDefs()[1];
+        arg_not_in_subgraph = cur->MutableInputDefs()[0];
+      }
+
+      // arg_in_subgraph is contained in candidate_node_args, so its shape must be [batch_size, seq_len, ...]
+      // Now only support cases of the two shapes are absolutely same or the other shape dim size is smaller by 2.
+      // For example, [batch_size, seqlen, hidden_size] and [batch_size, seqlen, hidden_size].
+      //              [batch_size, seqlen, hidden_size] and [seqlen].
+      // TODO: support other case such as:
+      //              [batch_size, seqlen, hidden_size] and [batch_size, 1, hidden_size]
+      if (arg_not_in_subgraph->Shape()->dim_size() <= arg_in_subgraph->Shape()->dim_size() - 2 ||
```

Comment:
be careful with arg_not_in_subgraph->Shape(), better add a null check before using it